File: R/getbasismatrix.R
getbasismatrix <- function(evalarg, basisobj, nderiv=0) {
#  Computes the basis matrix evaluated at arguments in EVALARG associated
#    with basis.fd object BASISOBJ.  The basis matrix contains the values
#    at argument value vector EVALARG of applying the nonhomogeneous
#    linear differential operator LFD to the basis functions.  By default
#    LFD is 0, and the basis functions are simply evaluated at argument
#    values in EVALARG.
#
#  If LFD is a functional data object with m + 1 functions c_1, ... c_{m+1}, then it
#    is assumed to define the order m HOMOGENEOUS linear differential operator
#  Lx(t) = c_1(t) + c_2(t)x(t) + c_3(t)Dx(t) + ... + c_{m+1}D^{m-1}x(t) + D^m x(t).
#
#  If the basis type is either polygonal or constant, LFD is ignored.
#
#  Arguments:
#  EVALARG ... Either a vector of values at which all functions are to evaluated,
#              or a matrix of values, with number of columns corresponding to
#              number of functions in argument FD.  If the number of evaluation
#              values varies from curve to curve, pad out unwanted positions in
#              each column with NA.  The number of rows is equal to the maximum
#              of number of evaluation points.
#  BASISOBJ ... A basis object
#  NDERIV   ... A nonnegative integer indicating a derivative to be evaluated.

#
#  Note that the first two arguments may be interchanged.
#
#  Last modified 26 October 2005

#  Exchange the first two arguments if the first is an BASIS.FD object
#    and the second numeric

if (is.numeric(basisobj) && inherits(evalarg, "basisfd")) {
    temp     <- basisobj
    basisobj <- evalarg
    evalarg  <- temp
}

#  check EVALARG

if (!(is.numeric(evalarg)))  stop("Argument EVALARG is not numeric.")
	
#  check basisobj
	
if (!(inherits(basisobj, "basisfd"))) stop(
    "Second argument is not a basis object.")

#  Extract information about the basis

type     <- basisobj$type
nbasis   <- basisobj$nbasis
params   <- basisobj$params
rangeval <- basisobj$rangeval
dropind  <- basisobj$dropind

#  -----------------------------  B-spline basis  -------------------

if (type == "bspline") {
   	breaks   <- c(rangeval[1], params, rangeval[2])
   	norder   <- nbasis - length(breaks) + 2
   	basismat <- bsplineS(evalarg, breaks, norder, nderiv)

#  -----------------------------  Constant basis  --------------------

} else if (type == "const") {
   	basismat  <- matrix(1,length(evalarg),1)

#  -----------------------------  Exponential basis  -------------------

} else if (type == "expon") {
   	basismat  <- expon(evalarg, params, nderiv)

#  -------------------------------  Fourier basis  -------------------

} else if (type == "fourier") {
   	period   <- params[1]
   	basismat <- fourier(evalarg, nbasis, period, nderiv)

#  -----------------------------  Monomial basis  -------------------

} else if (type == "monom") {
   	basismat  <- monomial(evalarg, params, nderiv)

#  -----------------------------  Polygonal basis  -------------------

} else if (type == "polyg") {
    basismat  <- polyg(evalarg, params)

#  -----------------------------  Polynomial basis  -------------------

} else if (type == "polynom") {
   	norder   <- nbasis
   	ctr      <- params[1]
   	basismat <- polynom(evalarg, norder, nderiv, ctr)

#  -----------------------------  Power basis  -------------------

} else if (type == "power") {
    basismat  <- powerbasis(evalarg, params, nderiv)

#  -----------------------  Unrecognizable basis  --------------------

} else {
   	stop("Basis type not recognizable")
}
	
#  remove columns for bases to be dropped

if (length(dropind) > 0) basismat <- basismat[,-dropind]

return(basismat)

}

File: inst/Matlab/fdaM/@basis/getbasismatrix.m
function basismat = getbasismatrix(evalarg, basisobj, nderiv)
%  GETBASISMATRIX   Computes the basis matrix evaluated at arguments in
%    EVALARG associated with basis.fd object BASISOBJ.
%    The returned basis matrix BASISMAT contains the basis
%    derivatives of order NDERIV (0 by default).

%  last modified 7 March 2011 by Jim Ramsay

if nargin < 3,  nderiv = 0;  end

%  check basisobj

if ~isa_basis(basisobj)
    error('Argument BASISOBJ is not a functional basis object');
end

%  search for stored basis matrix

basismat = getbasisvalues(basisobj, evalarg, nderiv);

if ~isempty(basismat)

    %  stored basis matrix found

    return;

else

    %  evaluate basismatrix

    type     = getbasistype(basisobj);
    nbasis   = getnbasis(basisobj);
    params   = getbasispar(basisobj);
    rangeval = getbasisrange(basisobj);
    dropind  = getdropind(basisobj);

    switch type
        case 'bspline'
            rangex   = rangeval;
            if isempty(params)
                breaks = rangex;
            else
                breaks   = [rangex(1), params, rangex(2)];
            end
            norder   = nbasis - length(breaks) + 2;
            basismat = bsplineM(evalarg, breaks, norder, nderiv);
        case 'fourier'
            period   = params(1);
            basismat = fourier(evalarg, nbasis, period, nderiv);
        case 'monom'
            basismat = monomial(evalarg, params, nderiv);
        case 'polyg'
            basismat = polyg(evalarg, params);
        case 'power'
            basismat = powerbasis(evalarg, params, nderiv);
        case 'expon'
            exponents = params;
            basismat = expon(evalarg, exponents, nderiv);
        case 'const'
            basismat = ones(length(evalarg),1);
        case 'QW'
            basismat = QW(evalarg, nderiv);
        case 'QWM'
            basismat = QWM(evalarg, nderiv);
        case 'QS'
            basismat = QS(evalarg, nbasis, nderiv);
        case 'slide'
            rangex   = rangeval;
            if isempty(params)
                breaks = rangex;
            else
                breaks = [rangex(1), params(1:(nbasis-1)), rangex(2)];
                rates  = params(nbasis:(2*nbasis-1));
            end
            basismat = slides(evalarg, breaks, rates, nderiv);
        case 'fd'
            basismat = eval_fd(evalarg, params, nderiv);
        case 'FEM'
            nodes    = params.nodes;
            nodemesh = params.nodemesh;
            order    = params.order;
            basismat = FEM(evalarg, nodes, nodemesh, order, nderiv);
        case 'TP'
            error('GETBASISMATRIX not implemented for TP basis objects');
        case 'fdVariance'
            T = max(getbasisrange(basisobj));
            pars = getbasispar(basisobj);
            I = pars.I;
            J = pars.J;
            delta = T./I;
            B = delta.*J;
            basismat = RstCellSetup(evalarg, T, B, delta);
        otherwise
            error('Basis type not recognizable');
    end

    if ~isempty(dropind)
        index = 1:nbasis;
        for i=1:length(dropind)
            index = index(index ~= dropind(i));
        end
        basismat = basismat(:,index);
    end

end